### ISSUE TITLE
Add application-level connection locking

### ISSUE THREAD

**Issue by AnyhowStep** [AnyhowStep/tsql#55]
Given the following code,
```ts
connection.transaction(async (connection) => {
  await Promise.all([
    tableA.insertAndFetch(connection, dataA);
    tableB.insertAndFetch(connection, dataB);
  ]);
});
```

The following can happen,
1. Promise 1: dataA is inserted to tableA
1. Promise 2: dataB is inserted to tableB
1. Promise 1: `LAST_INSERT_ID()` is used; gets tableB's `LAST_INSERT_ID()`
1. Promise 2: `LAST_INSERT_ID()` is used; gets tableB's `LAST_INSERT_ID()`
1. Promise 1: Fetches wrong row
1. Promise 2: Fetches correct row

Variants of the above can happen whenever a function uses a connection to perform multiple operations... And we decide to invoke them all in parallel.

The best way to prevent this problem is to just `await` every operation that uses `connection`. That is, make sure all operations are sequential and not parallel.

I personally do this by having a lint rule for `no-floating-promises` and never using `Promise.all()`.

But we might be able to do better for people who do not have such rules enabled.

-----

For example, if a function wants to perform multiple operations using `connection`, it should do something like,
```ts
const promise1 = connection.lock(async (connection) => {
  await //snip insertOne()
  await //snip get LAST_INSERT_ID()
  return //snip fetchOne() using LAST_INSERT_ID()
});
//This will wait for promise1's lock to resolve before running
const promise2 = tableB.insertAndFetch(connection, dataB);
```

You can also call `.lock()` repeatedly
```ts
const promise1 = connection.lock(async (connection) => {
  await //snip insertOne()
  await //snip get LAST_INSERT_ID()
  return //snip fetchOne() using LAST_INSERT_ID()
});
//This will wait for promise1's lock to resolve before running
const promise2 = connection.lock(async (connection) => {
  await //snip insertOne()
  await //snip get LAST_INSERT_ID()
  return //snip fetchOne() using LAST_INSERT_ID()
});
//Looks like it'll run in arbitrary order, but will actually be sequential
await Promise.all([
  promise1,
  promise2,
]);
```

-----

### Nested `.lock()`

If `.lock()` is called on a locked connection, the lock is nested an extra layer.
```ts
connectionA.lock(async (connectionB) => {
  console.log(connectionA === connectionB); //false

  //OK
  const promiseB1 = tableB.insertAndFetch(connectionB, dataB);

  const promiseB2 = connectionB.lock(async (connectionC) => {
    console.log(connectionA === connectionB); //false
    console.log(connectionB === connectionC); //false

    
  });

  //Waits for promiseB2's lock to resolve before running
  const promiseB3 = tableB.insertAndFetch(connectionB, dataB);

  //After promiseB2 resolves, we know promiseB3 will execute
  await promiseB2;

  //Runs immediately because connectionB's nested lock has resolved
  const promiseB4 = tableB.insertAndFetch(connectionB, dataB);
});
```

-----

### Functions executing only one SQL statement

Some library functions only execute one SQL statement.
They may or may not call `.lock()`.
So, it is still possible to have SQL statements run in arbitrary order.

For example,
```ts
const results = await Promise.all([
  tableA.insertOne(connection, dataA),
  from(tableA).select(columns => [columns]).fetchAll(connection),
]);
```

It is possible for the `SELECT` statement to run before the `INSERT` statement and not throw an error.

However, the data returned by `insertOne()` is consistent within its own function body.
And the data returned by `fetchAll()` is consistent within its own function body.

The idea behind `.lock()` is to just enforce correctness within a function's own body.
Not across different function calls.

-----

Pros:
+ Prevents errors caused by making a single connection do too many different things at once, with an undefined order

Cons:
+ Additional overhead to create locks, and check them
+ If always `await`ing `connection` operations and never using `Promise.all()`, you will never run into this problem (extra work is done for no benefit)

-----

@G-Rath What do you think of this? Also, I'm not sure if "lock" is the best word for this. Or maybe it is. I can't seem to think of anything better.

**Comment by AnyhowStep** [comment]
Basically, we want to avoid this problem,
```ts
async function foo () {
  	await console.log("A")
	await console.log("B")
}
foo();
foo();
```

Where the output is,
```
A
A
B
B
```

When we really want,
```
A
B
A
B
```

**Comment by G-Rath** [comment]
> If always awaiting connection operations and never using Promise.all(), you will never run into this problem (extra work is done for no benefit)

That's super painful from a UX POV.

> Also, I'm not sure if "lock" is the best word for this

```
connection.doInSync
connection.sync
```

-------

> //Looks like it'll run in arbitrary order, but will actually be sequential
> await Promise.all([
>   promise1,
>   promise2,
> ]);

**Be very careful with this**

`Promise.all` is there for running things in parallel, so you shouldn't fight the native api - instead users should be educated to understand the consequences of using that API (Since there are valid uses for it I'm sure).

>Some library functions only execute one SQL statement.
They may or may not call .lock().
So, it is still possible to have SQL statements run in arbitrary order.

> For example,

That's a perfect example of where this is on the user, not the library :)

However, the example you gave about `insertAndFetch` to me is a reason to make `Connection` self-locking in some cases, since since you'll *always* have this problem regardless of if you're using `Promise.all`.

For example, what if I had two api requests for the same endpoint come in? If I'm the size of Facebook or Google, I'm serving enough requests at speed that I could reuse a connection in my pool every so often.

Hence I think this makes sense:

```
class Connection {
  private _insertLock = Promise.resolve();

  insertAndFetch(/* ... */) {
    return (this._insertLock = this._insertLock.then(() => {
       /* ... */
    }));
  }
}
```

That's a low-cost abstraction that at worst won't break anything, while at best preventing massive database problems.

------

> I personally do this by having a lint rule for no-floating-promises and never using Promise.all().

I wouldn't be surprised if it made sense to make a small eslint plugin for this library, given what it's doing - I imagine there's a number of pitfalls & de-ops that are best practice & hygiene based, of which this sounds like one.

Given my work on `eslint-plugin-jest`, I'd be happy to be a core contributor :)

**Comment by AnyhowStep** [comment]
Different requests use different connections, allocated and deallocated from a pool.

The pool should not return a connection instance that hasn't been released. And users should be using stuff like reading connection state before setting them (stuff like last_insert_id())

**Comment by G-Rath** [comment]
> Multiple connections running in parallel is generally fine because I assume people understand SQL transactions, and how to use them as appropriate.

Then why is that not the case for `Promise.all`?

> But other operations like insertAndFetch, updateAndFetch, Log.track, etc. Should be expected to work, even when called in parallel.

Which is why I think the self-locking example I gave might be the way to go, as a `lock` method only shifts the expectation around: You're now expecting users to know when a function needs to be locked if it's being done in parallel :)

**Comment by AnyhowStep** [comment]
Sorry, I just fleshed out my response above and edited the comment. I tend to comment before I'm done thinking things through =x

-----

> > Multiple connections running in parallel is generally fine because I assume people understand SQL transactions, and how to use them as appropriate.
> 
> Then why is that not the case for `Promise.all`?

It's fine to run multiple connections in parallel because there are ways for people to ensure connectionA does not affect the results of connectionB (using database transactions).
 
However, SQL has no concept of preventing "async task A" from modifying the connection session state of "async task B". If both tasks use the same connection, then they share the same session state.

Changes made by A affect B and vice versa. We don't have "session state transactions" or anything similar because SQL isn't concerned about that. So, we need a way to prevent A and B from messing each other up using some application-level concept. (Like `.lock()/.sync()`)

So, I guess the real motivation behind this isn't so much "We don't want multiple tasks to run SQL statements in parallel".

It's more like, "We don't want multiple tasks to read and write to connection session state in parallel".

-----

> > But other operations like insertAndFetch, updateAndFetch, Log.track, etc. Should be expected to work, even when called in parallel.
> 
> Which is why I think the self-locking example I gave might be the way to go, as a `lock` method only shifts the expectation around: You're now expecting users to know when a function needs to be locked if it's being done in parallel :)

With `.lock()` implemented, they won't need to call `.lock()` before calling the library's built-in functions/methods. I'm not shifting the expectation down to them.

I'll be using `.lock()` internally (I'll try my best to not forget, I promise) to prevent these problems with my stuff. I'll be using `.lock()` to self-lock, basically.

So, `Promise.all([ tableA.insertAndFetch(), tableB.insertAndFetch() ])` can be expected to work fine without calling `.lock()` at all.

-----

However, I still want to expose `.lock()` because downstream users may want to write their own abstractions that may use lower level SQL statements and touch connection session state directly.

In these situations, it would be nice to have something built-in that they can leverage to protect themselves.

**Comment by AnyhowStep** [comment]
Oh, of course, if they do something like,
```ts
Promise.all([ tableA.insertAndFetch(x), tableA.insertAndFetch(y) ])
```

It's possible for `y` to insert first. Then, `x`.

But that falls under,
> If it's "obvious", then it's your fault.

**Comment by G-Rath** [comment]
> Sorry, I just fleshed out my response above and edited the comment. I tend to comment before I'm done thinking things through =x

hehe all good - I do the same 😄 
--------

> Somewhat related but, basically, connection objects are not "thread safe" because of that session state. If you only ever use a connection in synchronous code (like with a language like PHP), you'll never run into these problems.

> I'll be using .lock() internally (I'll try my best to not forget, I promise) to prevent these problems with my stuff. I'll be using .lock() to self-lock, basically.

Ah now I'm with you - we're actually talking about the same thing, just looking at it from a slightly different angle - `lock` is the generic implementation of the self-locking I was talking about 👍

The key part you left out in the original issue body was "I'm going to be the one calling lock in most cases" :P

**Comment by AnyhowStep** [comment]
I played around with the locking thing and it worked really well.

I wrote a few `Promise.all()` tests and watched them all fail (of course, I picked cases where I knew it would fail).

Then, I added a simple lock implementation and it all passed.

-----

Turns out, I didn't have to do a lot of work to integrate the locking mechanism because of transactions.

I was already using transactions where ever necessary. And, turns out, transactions are also where you would want a lock.

It would be bad if taskA used taskB's transaction by accident!

So, I made acquiring a transaction also acquire a connection lock.

-----

I'll clean up the code and make a PR some time tonight
